python/pyarrow/parquet.py

```diff
@@ -807,6 +807,80 @@ def write_table(table, where, row_group_size=None, version='1.0',
         writer.close()
 
 
+def write_to_dataset(table, where, partition_cols, root_path,
+                     preserve_index=True, **kwargs):
+    """
+    Wrapper around parquet.write_table for writing a Table to
+    Parquet format by partitions.
+    For each combination of partition columns and values,
+    a subdirectories are created in the following
+    manner:
+
+    root_dir/
+    ├── group1=value1
+    │   ├── group2=value1
+    │   │   └── <result.parquet>
+    │   └── group2=value2
+    │       └── <result.parquet>
+    ├── group1=value2
+    │   └── group2=value1
+    │       └── <result.parquet>
+    ...
+    └── group1=valueN
+        ├── group2=value1
+        │   └── <result.parquet>
+        ...
+        └── group2=valueN
+            └──<result.parquet>
+
+    Parameters
+    ----------
+    table : pyarrow.Table
+    where: string,
+        Name of the parquet file for data saved in each partition
+    parition_cols : list,
+        Column names by which to partition the dataset
+        Columns are partitioned in the order they are given
+    root_path: string,
+        The root directory of the table
+    **kwargs : dict, kwargs for write_table function.
+    """
+    from pyarrow import Table
+
+    df = table.to_pandas()
+    groups = df.groupby(partition_cols)
+    data_cols = [col for col in df.columns.tolist()
+                 if col not in partition_cols]
+    for partition in partition_cols:
+        try:
+            df[partition] = df[partition].astype(str)
+        except ValueError:
+            raise ValueError("Partition column must be coercible to string")
+
+    if not data_cols:
+        raise ValueError("No data left to save outside partition columns")
+
+    schema = {}
+    for subgroup in groups.indices:
+        sub_df = groups.get_group(subgroup)[data_cols]
```

Comment:
Instead do `for keys, subgroup in df.groupby(...)` and omit the `.indices` and `.get_group`

Reply:
I think it would be more efficient to do:

```python
partition_keys = [df[col] for col in partition_cols]
data_df = df.drop(partition_cols, axis='columns')
for part_keys, data_group in data_df.groupby(partition_keys):
    ...
```